packages/engine/lib/error/src/lib.rs

```diff
@@ -0,0 +1,510 @@
+///
+/// # Examples
+///
+/// `Result` can also be used in `fn main()`:
+///
+/// ```
+/// # fn has_permission(_: usize, _: usize) -> bool { true }
+/// # fn get_user() -> Result<usize> { Ok(0) }
+/// # fn get_resource() -> Result<usize> { Ok(0) }
+/// use error::{ensure, Result};
+///
+/// fn main() -> Result<()> {
+///     let user = get_user()?;
+///     let resource = get_resource()?;
+///
+///     ensure!(
+///         has_permission(user, resource),
+///         "Permission denied for {user} accessing {resource}"
+///     );
+///
+///     //...
+///     # Ok(())
+/// }
+/// ```
+pub type Result<T, Context = ()> = core::result::Result<T, Report<Context>>;
+
+/// Extension trait for [`Result`][core::result::Result] to provide context information on
+/// [`Report`]s.
+pub trait ResultExt<T> {
+    /// Type of the resulting context `C` inside of [`Report<C>`] when not providing a context.
+    type Context;
+
+    /// Adds new contextual message to the [`Frame`] stack of a [`Report`].
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use error::Result;
+    /// # fn load_resource(_: &User, _: &Resource) -> Result<()> { Ok(()) }
+    /// # struct User;
+    /// # struct Resource;
+    /// use error::ResultExt;
+    ///
+    /// # let user = User;
+    /// # let resource = Resource;
+    /// # #[allow(unused_variables)]
+    /// let resource = load_resource(&user, &resource).wrap_err("Could not load resource")?;
+    /// # Result::Ok(())
+    /// ```
+    fn wrap_err<M>(self, message: M) -> Result<T, Self::Context>
+    where
+        M: fmt::Display + fmt::Debug + Send + Sync + 'static;
+
+    /// Lazily adds new contextual message to the [`Frame`] stack of a [`Report`].
+    ///
+    /// The function is only executed in the `Err` arm.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use core::fmt;
+    /// # use error::Result;
+    /// # fn load_resource(_: &User, _: &Resource) -> Result<()> { Ok(()) }
+    /// # struct User;
+    /// # struct Resource;
+    /// # impl fmt::Display for Resource { fn fmt(&self, _: &mut fmt::Formatter<'_>) -> fmt::Result { Ok(()) }}
+    /// use error::ResultExt;
+    ///
+    /// # let user = User;
+    /// # let resource = Resource;
+    /// # #[allow(unused_variables)]
+    /// let resource = load_resource(&user, &resource)
+    ///     .wrap_err_lazy(|| format!("Could not load resource {resource}"))?;
+    /// # Result::Ok(())
+    /// ```
+    fn wrap_err_lazy<M, F>(self, op: F) -> Result<T, Self::Context>
+    where
+        M: fmt::Display + fmt::Debug + Send + Sync + 'static,
+        F: FnOnce() -> M;
+
+    /// Adds a context provider to the [`Frame`] stack of a [`Report`] returning
+    /// [`Result<T, Context>`]).
+    // TODO: come up with a decent example
+    fn provide_context<C>(self, context: C) -> Result<T, C>
+    where
+        C: Provider + fmt::Display + fmt::Debug + Send + Sync + 'static;
```

Comment:
Again, something stopping us using the `Context` trait on these?

Reply:
Not possible here as it's part of the public API

Reply:
Can we not expose `Context` publicly?

Reply:
Have discussed off GH and it's been argued that doing so wouldn't improve anything for the user and would likely just serve to confuse them if the trait is empty